### PR TITLE
Native hamming distance

### LINKED ISSUE #3: Add native implementations of algorithms
This library is currently just a wrapper around `org.apache.commons.commons-text` functions.

Native implementations would be a lot better... Who is ready for a hard Spark coding challenge?!

@afranzi @nvander1 - take a look at the README for this project.  Want to get this on your radar 🤓

### PR DESCRIPTION
**Pull request by nvander1**
Still a WIP, don't merge yet!

Fixes #3 

@MrPowers Here is a first draft of a native HammingDistance

I need to find out whether the UTF8String is getting copied on the call to getBytes(). Sadly the getByte(int) method on UTF8String is private, otherwise I'd prefer to use that. I'm going to look into it to see if there's a way to get access. Maybe reflection, but that also has drawbacks.

Also, since the apache commons library creates a lot of intermediate objects ( ie `new CosineDistance()`), it will likely be faster to implement the algorithms directly on the UTF8String that spark provides. This is the approach Spark takes on the built-in levenshtein function. See: https://github.com/apache/spark/blob/v2.4.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L1656-L1670

https://github.com/apache/spark/blob/master/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L1278-L1340

### CODE REVIEW

**Comment by nvander1** [comment]
FYI @MrPowers,  @ethanhoman  was disuccsing the need for support of various apache-commons-text functions at work before you approached me about this project. Seems great minds think alike! :smile_cat:

File: src/main/scala/com/github/mrpowers/spark/stringmetric/functions.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.github.mrpowers.spark.stringmetric
+
+import com.github.mrpowers.spark.stringmetric.expressions._
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.catalyst.expressions.Expression
+
+object functions {
```
Comment (nvander1) [inline src/main/scala/com/github/mrpowers/spark/stringmetric/functions.scala]: I put it in a separate functions object for now @MrPowers. I can continue this way to be consistent with the spark.sql.functions name convention, or just replace the one in SimilartyFunctions if you prefer not to break the api.

**Comment by MrPowers** [comment]
@nvander1 - This PR is off to a great start!

A wrote [a blog post on the different types of Spark functions](https://medium.com/@mrpowers/the-different-type-of-spark-functions-custom-transformations-column-functions-udfs-bf556c9d0ce7) and [Sim pointed out that there are also Spark native functions](https://medium.com/@sim_46373/great-post-3b135ad6c190).

Here is [Sim's article on Spark native functions](https://blog.simeonov.com/2018/11/14/apache-spark-native-functions/).

Is your implementation here similar to the Spark native functions Sim described?

**Comment by nvander1** [comment]
Yes it's how the builtin string functions like levensthein are implemented.

**Comment by nvander1** [comment]
@MrPowers 

You can see the actual java code generated for the hamming test here: (I just added a quick println in the test `println(actualDF.queryExecution.debug.codegen)`
The line where the select is happening is line 31 of the generated code.

```
...
[info] hamming
Found 1 WholeStageCodegen subtrees.
== Subtree 1 / 1 ==
*(1) Project [word1#84, word2#85, hamming(word1#84, word2#85) AS w1_w2_hamming#88]
+- Scan ExistingRDD[word1#84,word2#85]

Generated code:
/* 001 */ public Object generate(Object[] references) {
/* 002 */   return new GeneratedIteratorForCodegenStage1(references);
/* 003 */ }
/* 004 */
/* 005 */ // codegenStageId=1
/* 006 */ final class GeneratedIteratorForCodegenStage1 extends org.apache.spark.sql.execution.BufferedRowIterator {
/* 007 */   private Object[] references;
/* 008 */   private scala.collection.Iterator[] inputs;
/* 009 */   private scala.collection.Iterator inputadapter_input_0;
/* 010 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[] project_mutableStateArray_0 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[1];
/* 011 */
/* 012 */   public GeneratedIteratorForCodegenStage1(Object[] references) {
/* 013 */     this.references = references;
/* 014 */   }
/* 015 */
/* 016 */   public void init(int index, scala.collection.Iterator[] inputs) {
/* 017 */     partitionIndex = index;
/* 018 */     this.inputs = inputs;
/* 019 */     inputadapter_input_0 = inputs[0];
/* 020 */     project_mutableStateArray_0[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(3, 64);
/* 021 */
/* 022 */   }
/* 023 */
/* 024 */   private void project_doConsume_0(InternalRow inputadapter_row_0, UTF8String project_expr_0_0, boolean project_exprIsNull_0_0, UTF8String project_expr_1_0, boolean project_exprIsNull_1_0) throws java.io.IOException {
/* 025 */     boolean project_isNull_2 = true;
/* 026 */     int project_value_2 = -1;
/* 027 */
/* 028 */     if (!project_exprIsNull_0_0) {
/* 029 */       if (!project_exprIsNull_1_0) {
/* 030 */         project_isNull_2 = false; // resultCode could change nullability.
/* 031 */         project_value_2 = com.github.mrpowers.spark.stringmetric.unsafe.UTF8StringFunctions.hammingDistance(project_expr_0_0, project_expr_1_0);
/* 032 */
/* 033 */       }
/* 034 */
/* 035 */     }
/* 036 */     project_mutableStateArray_0[0].reset();
/* 037 */
/* 038 */     project_mutableStateArray_0[0].zeroOutNullBytes();
/* 039 */
/* 040 */     if (project_exprIsNull_0_0) {
/* 041 */       project_mutableStateArray_0[0].setNullAt(0);
/* 042 */     } else {
/* 043 */       project_mutableStateArray_0[0].write(0, project_expr_0_0);
/* 044 */     }
/* 045 */
/* 046 */     if (project_exprIsNull_1_0) {
/* 047 */       project_mutableStateArray_0[0].setNullAt(1);
/* 048 */     } else {
/* 049 */       project_mutableStateArray_0[0].write(1, project_expr_1_0);
/* 050 */     }
/* 051 */
/* 052 */     if (project_isNull_2) {
/* 053 */       project_mutableStateArray_0[0].setNullAt(2);
/* 054 */     } else {
/* 055 */       project_mutableStateArray_0[0].write(2, project_value_2);
/* 056 */     }
/* 057 */     append((project_mutableStateArray_0[0].getRow()));
/* 058 */
/* 059 */   }
/* 060 */
/* 061 */   protected void processNext() throws java.io.IOException {
/* 062 */     while (inputadapter_input_0.hasNext() && !stopEarly()) {
/* 063 */       InternalRow inputadapter_row_0 = (InternalRow) inputadapter_input_0.next();
/* 064 */       boolean inputadapter_isNull_0 = inputadapter_row_0.isNullAt(0);
/* 065 */       UTF8String inputadapter_value_0 = inputadapter_isNull_0 ?
/* 066 */       null : (inputadapter_row_0.getUTF8String(0));
/* 067 */       boolean inputadapter_isNull_1 = inputadapter_row_0.isNullAt(1);
/* 068 */       UTF8String inputadapter_value_1 = inputadapter_isNull_1 ?
/* 069 */       null : (inputadapter_row_0.getUTF8String(1));
/* 070 */
/* 071 */       project_doConsume_0(inputadapter_row_0, inputadapter_value_0, inputadapter_isNull_0, inputadapter_value_1, inputadapter_isNull_1);
/* 072 */       if (shouldStop()) return;
/* 073 */     }
/* 074 */   }
/* 075 */
/* 076 */ }


()
[info] - computes the hamming metric
```

**Comment by MrPowers** [comment]
@nvander1 - I migrated this project to Mill, so this rebase might be a bit annoying... sorry about that!

Let me know if you'd like me to rebase this branch and then pass it back off to you.

**Comment by nvander1** [comment]
@MrPowers Rebased your changes into the branch. I think its good to go for hamming. I'll start on the other functions in different PRs. cc: @manuzhang

**Comment by nvander1** [comment]
I was wrong, this is not ready. This will report difference at the byte level not char level.

**Comment by nvander1** [comment]
I need to do more work to loop over the unicode characters correctly. The newest test case highlights this bug.

**Comment by nvander1** [comment]
Basic multi byte hamming tests now pass. I want to add some more complex test cases for my own peace of mind.

This is gonna be fun+challenging when we get to other native string functions.

**Comment by nvander1** [comment]
@MrPowers Actually ready to go now. @manuzhang

**Comment by MrPowers** [comment]
Great work @nvander1!  Great catch on making this work for multi-byte characters.

Feel free to keep the PRs coming and I'll make sure they get merged.

Let's chat about next steps.  I think we should benchmark this, quantify gains, and then write a blog post.

Then we should try to get this in the Spark core library / tell the Spark team to make more stuff public.

**Pull request merged**